capsules/src/sdcard.rs

```diff
@@ -0,0 +1,1424 @@
+            write_buffer[2] = 0x40 | cmd as u8;
+        }
+
+        // argument, MSB first
+        write_buffer[3] = ((arg >> 24) & 0xFF) as u8;
+        write_buffer[4] = ((arg >> 16) & 0xFF) as u8;
+        write_buffer[5] = ((arg >> 8) & 0xFF) as u8;
+        write_buffer[6] = ((arg >> 0) & 0xFF) as u8;
+
+        // CRC is ignored except for CMD0 and maybe CMD8
+        if cmd == SDCmd::CMD8 {
+            write_buffer[7] = 0x87; // valid crc for CMD8(0x1AA)
+        } else {
+            write_buffer[7] = 0x95; // valid crc for CMD0
+        }
+
+        // append dummy bytes to transmission
+        for i in 0..recv_len {
+            write_buffer[8 + i] = 0xFF;
+        }
+
+        self.spi.read_write_bytes(write_buffer, Some(read_buffer), 8 + recv_len);
+    }
+
+    fn read_bytes(&self,
+                  mut write_buffer: &'static mut [u8],
+                  mut read_buffer: &'static mut [u8],
+                  recv_len: usize) {
+
+        self.set_spi_fast_mode();
+
+        // set write buffer to null transactions
+        // Note: this could be optimized in the future by allowing SPI to read
+        //  without a write buffer passed in
+        let count = cmp::min(write_buffer.len(), recv_len);
+        for i in 0..count {
+            write_buffer[i] = 0xFF;
+        }
+
+        self.spi.read_write_bytes(write_buffer, Some(read_buffer), recv_len);
+    }
+
+    fn write_bytes(&self,
+                   mut write_buffer: &'static mut [u8],
+                   mut read_buffer: &'static mut [u8],
+                   recv_len: usize) {
+
+        self.set_spi_fast_mode();
+
+        self.spi.read_write_bytes(write_buffer, Some(read_buffer), recv_len);
+    }
+
+
+    fn get_response(&self, response: SDResponse, read_buffer: &[u8]) -> (u8, u8, u32) {
+
+        let mut r1: u8 = 0xFF;
+        let mut r2: u8 = 0xFF;
+        let mut r3: u32 = 0xFFFFFFFF;
+
+        // scan through read buffer for response byte
+        for i in 0..read_buffer.len() {
+            if (read_buffer[i] & 0x80) == 0x00 {
+                // status byte is always included
+                r1 = read_buffer[i];
+
+                match response {
+                    SDResponse::R2 => {
+                        // status, then read/write status. Unused in practice
+                        if i + 1 < read_buffer.len() {
+                            r2 = read_buffer[i + 1];
+                        }
+                    }
+                    SDResponse::R3 | SDResponse::R7 => {
+                        // status, then Operating Condition Register
+                        if i + 4 < read_buffer.len() {
+                            r3 = (read_buffer[i + 1] as u32) << 24 |
+                                 (read_buffer[i + 2] as u32) << 16 |
+                                 (read_buffer[i + 3] as u32) << 8 |
+                                 (read_buffer[i + 4] as u32);
+                        }
+                    }
+                    _ => {}
```

Comment:
Are these unexpected / unhandled responses, or things you're deliberately ignoring? If the former, maybe put a `debug!("Unexpected SDResponse: {}", response)` or some such in the body?